1. Server-Side Scenarios/msal-client-credential-secret-high-availability/src/main/java/ClientCredentialGrant.java

```diff
@@ -0,0 +1,130 @@
+import java.util.Properties;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+class ClientCredentialGrant {
+
+    private static String instance;
+    private static String tenantId;
+    private static String clientId;
+    private static String secret;
+    private static String scope;
+
+
+    public static void main(String args[]) throws Exception{
+
+        Properties properties = new Properties();
+        properties.load(Thread.currentThread().getContextClassLoader().getResourceAsStream("application.properties"));
+        instance = properties.getProperty("INSTANCE");
+        clientId = properties.getProperty("CLIENT_ID");
+        secret = properties.getProperty("SECRET");
+
+        // With client credentials flows the scope is ALWAYS of the shape "resource/.default", as the
+        // application permissions need to be set statically (in the portal), and then granted by a tenant administrator
+        scope = properties.getProperty("SCOPE");
+
+        // A multi-tenanted app can have a list of these. There will be 1 token per tenant.
+        // Multi-tenanted services might feature millions of tenants so a token cache with size eviction is needed
+        // to avoid out of memory issues
+        String tenantId = properties.getProperty("TENANT_ID");
+
+
+        try {
+
+            IAuthenticationResult result = getAccessTokenByClientCredentialGrant(instance, clientId, tenantId, secret, scope);
+
+            // If you try to fetch the same token again, it will hit the token cache and it will be much faster.
+            // In case AAD has an outage, cached tokens are still available, increasing your app's resiliency.
+            result = getAccessTokenByClientCredentialGrant(instance, clientId, tenantId, secret, scope);
+
+
+
+            String usersListFromGraph = getUsersListFromGraph(result.accessToken());
+            System.out.println("Users in the Tenant = " + usersListFromGraph);
+            System.out.println("Press any key to exit ...");
+            System.in.read();
+
+        } catch(Exception ex){
+            System.out.println("Oops! We have an exception of type - " + ex.getClass());
+            System.out.println("Exception message - " + ex.getMessage());
+            throw ex;
+        }
+    }
+
+    private static IAuthenticationResult getAccessTokenByClientCredentialGrant(String instance, String clientId, String tenantId, String secret, String scope) throws Exception {
+
+        // Setup the token cache first. It is configured to allow 100k token entries, which at 2-3KB per entry, will take under 500MB of memory
+        // Note that size calculations are approximate and depend on the JVM
+        String cacheKey = clientId + "_" + tenantId + "_AppTokenCache";
+        MemoryTokenCacheWithEviction memoryTokenCacheWithEviction = new MemoryTokenCacheWithEviction(cacheKey);
+
+        ConfidentialClientApplication cca = ConfidentialClientApplication.builder(
```

Comment:
What is the collection behind internal cache in Java? Is it just a flat dictionary with the same access token cache key as used to be in .NET?
Are there any other pre-build caches for Java or do the users have to write their own cache provider?

Reply:
It is based on Google Guava Cache, which is pretty much like C#'s MemoryCache

https://github.com/google/guava/wiki/CachesExplained

Reply:
I meant the internal structure - like in .NET we have the concurrent dictionary with partitions. I assume here it's a dictionary but not partitioned, so FindAT filters through all the tokens?

Will the Guava be default static cache option or will it be just a static dictionary without cache eviction? (@Avery-Dunn I think it's related to you work)

Reply:
Yes, the internal structure is similar to the Accessor in MSAL .NET but the ones before your refactoring. FindAT filters through all tokens.

No, Guava should not be the default static cache option, I think @Avery-Dunn will use a similar trick that MSAL.NET uses for static caching.